### PR TITLE
[MST-477] Add a video upload tool to upload videos to edX's video encoding pipeline

### PR DESCRIPTION
**Pull request by MichaelRoytman**
[MST-477](https://openedx.atlassian.net/browse/MST-477)

This adds a command line utility that uploads videos to edX's video encoding pipeline. Please see the README.rst file for further details.

### CODE REVIEW

File: src/cc2olx/tools/README.rst [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+-----------------
+
+The video upload tool uploads video files to edX's video encoding pipeline.
+It also generates an output CSV that can be used as input to the video xBlock conversion tool.
+
+Input
+-----
+The tool takes as command line arguments a ``course_id``, a ``directory``, and a ``csv``.
+
+The ``course_id`` argument is the ID of the course as it appears in Studio. For example, ``course-v1:edX+111222+111222``.
```
Comment (MichaelRoytman) [inline src/cc2olx/tools/README.rst:14]: I'm not sure if this is the canonical place to get the course ID. Is there a better place I can refer to?

File: src/cc2olx/tools/README.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    python src/cc2olx/tools/video_upload.py course-v1:edX+111222+111222 /Users/example/workspaces/videos/Users/example/workspaces/video-data.csv > logs.txt
+
+Use
+-----
+The video upload tool is a command-line Python 3 program.
+
+In order to make authenticated calls to Studio to generate signed URLs for uploading videos, you need to supply valid credentials.
+
+If you are an edX staff user, you can use the client credentials associated with an oAuth2 application registered in the desired environment. If you are not an edX staff user, you will have to go through the process for generating a client ID and client secret. You can use the `instructions <https://course-catalog-api-guide.readthedocs.io/en/latest/authentication/#getting-a-client-id-and-client-secret
```
Comment (MichaelRoytman) [inline src/cc2olx/tools/README.rst:53]: I'm not positive about this. Alan was handling authentication, so I need to look into this more to make sure these are the correct instructions.

File: src/cc2olx/tools/video_upload.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
+import argparse
+import csv
+import os
+from pathlib import Path
+
+import requests
+from requests.auth import AuthBase
+
+
+OAUTH_TOKEN_URL = 'https://courses.edx.org/oauth2/access_token'
```
Comment (MichaelRoytman) [inline src/cc2olx/tools/video_upload.py:10]: We wanted to avoid having a full fledged Django application, which is why these are not in settings.

File: src/cc2olx/tools/README.rst [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    python src/cc2olx/tools/video_upload.py course-v1:edX+111222+111222 /Users/example/workspaces/videos/Users/example/workspaces/video-data.csv > logs.txt
+
+Use
+-----
+The video upload tool is a command-line Python 3 program.
+
+In order to make authenticated calls to Studio to generate signed URLs for uploading videos, you need to supply valid credentials.
+
+If you are an edX staff user, you can use the client credentials associated with a cc2olx oAuth2 application registered in the desired environment. If you are not an edX staff user, you will have to go through the process for generating a client ID and client secret. You can use the `instructions <https://course-catalog-api-guide.readthedocs.io/en/latest/authentication/#getting-a-client-id-and-client-secret
```
Comment (schenedx) [inline src/cc2olx/tools/README.rst:53]: To me, this is worth to be its own title and section.
It was not clear to me you expected the client_id and client_secret to be in the OS environment variables list.
Can we also have client_id and client_key optionally provided from the command line arguments?

File: src/cc2olx/tools/video_upload.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    Returns:
+        * response: the response object from the POST API call
+    """
+    s = requests.Session()
+    s.auth = SuppliedJwtAuth(access_token)
+
+    try:
+        response = s.post(url, json=data)
+        response.raise_for_status()
+    except requests.exceptions.HTTPError as error:
```
Comment (schenedx) [inline src/cc2olx/tools/video_upload.py:97]: Nice exception handling with logging. I like these.

File: src/cc2olx/tools/video_upload.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+                    file_data = data['files'][0]
+
+                    if 'edx_video_id' in file_data:
+                        edx_video_id = file_data['edx_video_id']
+
+                    if 'upload_url' in file_data:
+                        upload_url = file_data['upload_url']
+
+                # upload video to presigned url if we have one
+                if upload_url:
```
Comment (schenedx) [inline src/cc2olx/tools/video_upload.py:234]: Can we do some error handling before this logic? Raise and print an error if upload_url is `None`. Then raise a warning if the `edx_video_id` is None but continue to do the upload?

File: src/cc2olx/tools/README.rst [hunk 7]
```diff
@@ -0,0 +1,61 @@
+Tools
```
Comment (schenedx) [inline src/cc2olx/tools/README.rst:1]: Can you link the read me here into the overall repo's readme so others can see these instruction from the repo's home page?

**Review by schenedx — CHANGES_REQUESTED** [review]
See my inline comments.

I thought you are building a different script to convert the GT supplied csv into a desired CSV format. I didn't see that in this PR. Is that in a different PR?

**Comment by MichaelRoytman** [comment]
@schenedx I do have a script that converts the CSV that was supplied by Georgia Tech for this particular course into a useable format. The agreement Alie and I came up with is that, moving forward, we'd tell partners that they need to supply CSVs in the format that we expect, so this conversion will no longer be necessary. The script I wrote is hardcoded to work just for the CSV I was given. I'm happy to put it into a gist, if you want look at it, but I don't think it will be useful for other courses. Does that make sense?

So, basically, there were two steps for me:

1. convert GT's CSV into a useable format
2. use the new CSV to upload videos to edX's video encoding pipeline.

I have done both 1 and 2, but I think only 2 needs to be included in this repository.

**Comment by schenedx** [comment]
@MichaelRoytman I understand your reasons. Please share your script to translate from GT's provided csv into the desired format. I worry if anyone other than you wants to run to upload for GT course while you are away, we'd need to redo the work for the format conversation. If you don't feel comfortable checking into the repo, please put that script in a gist and share it in this PR.

File: src/cc2olx/tools/README.rst [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+Otherwise, the file is identical to the file supplied as a command line argument.
+
+
+Logs
+----
+The tool will log to the console any errors that are encountered during execution.
+
+You can redirect the output to a file for later inspection.::
+
+    python src/cc2olx/tools/video_upload.py course-v1:edX+111222+111222 /Users/example/workspaces/videos/Users/example/workspaces/video-data.csv > logs.txt
```
Comment (MichaelRoytman) [inline src/cc2olx/tools/README.rst:45]: TODO: fix this comment

File: src/cc2olx/tools/README.rst (hunk 4 above)
Comment (MichaelRoytman) [reply src/cc2olx/tools/README.rst:53]: I can separate this section out into a separate section. What would be the benefit, though? Better visibility? I figure that the `Use` section is fairly short and the necessary information is there.

I decided against allowing credentials as command line arguments to avoid pasting a client secret in the user's terminal. My opinion is that it is risky to paste cleartext secrets in a terminal session. I figure that a little extra nuisance is worth the added little bit of security. What do you think?

Comment (schenedx) [reply src/cc2olx/tools/README.rst:53]: Ok, that is a consideration worth writing down here. I agree to leave my suggestion aside.
I believe the first user of this tool needs to figure out Auth info. This is the section they need to read and it would be important to highlight. Therefore my suggestion above. Not critical.

File: src/cc2olx/tools/video_upload.py (hunk 6 above)
Comment (MichaelRoytman) [reply src/cc2olx/tools/video_upload.py:234]: I can add error handling. 

But, I don't think we should upload a video without an `edx_video_id`. In fact, I don't see how we could receive an `upload_url` but not an `edx_video_id` in the response, as the `edx_video_id` is just a UUID that we generate [here](https://github.com/edx/edx-platform/blob/8866cc0e231b2bd4a161be75417c64e198f8d24f/cms/djangoapps/contentstore/views/videos.py#L754). The edx video ID is what will allow the second step of this tool to convert the Common Cartridge videos into video xBlocks.

I'd prefer just to print an error and skip the video if either the `upload_url` is missing or the `edx_video_id` is missing.

File: src/cc2olx/tools/README.rst (hunk 7 above)
Comment (MichaelRoytman) [reply src/cc2olx/tools/README.rst:1]: Yes, I will do that.

**Comment by MichaelRoytman** [comment]
@schenedx Here is the [script](https://gist.github.com/MichaelRoytman/e6086d98109fa708447f9d2d78e0b31f) I used to convert the CSV I received from Georgia Tech. into the format this tool needs.

File: src/cc2olx/tools/README.rst (hunk 4 above)
Comment (MichaelRoytman) [reply src/cc2olx/tools/README.rst:53]: That makes sense. I'll incorporate those changes

**Review by schenedx — APPROVED** [review]
👍
